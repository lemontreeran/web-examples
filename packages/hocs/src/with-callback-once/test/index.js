import React from 'react'
import { mount } from 'enzyme'

import withCallbackOnce from '../src/'

const Target = () => null

describe('withCallbackOnce', () => {
  it('should pass props through', () => {
    const EnhancedTarget = withCallbackOnce()(Target)
    const wrapper = mount(
      <EnhancedTarget a={1} b={2} c={3} />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should invoke a callback when condition status has been changed', () => {
    const shouldCall = ({ a }) => a === 11
    const mockCallback = jest.fn()
    const EnhancedTarget = withCallbackOnce(shouldCall, mockCallback)(Target)
    const wrapper = mount(
      <EnhancedTarget a={1} b={2} c={3} />
    )

    expect(mockCallback.mock.calls).toMatchSnapshot()
    wrapper.setProps({ a: 11 })
    expect(mockCallback.mock.calls).toMatchSnapshot()
  })

  it('should no-op if condition status is the same', () => {
    const shouldCall = ({ b }) => b === 3
    const mockCallback = jest.fn()
    const EnhancedTarget = withCallbackOnce(shouldCall, mockCallback)(Target)
    const wrapper = mount(
      <EnhancedTarget a={1} b={2} c={3} />
    )

    expect(mockCallback.mock.calls).toMatchSnapshot()
    wrapper.setProps({ a: 11 })
    expect(mockCallback.mock.calls).toMatchSnapshot()
  })

  describe('display name', () => {
    const origNodeEnv = process.env.NODE_ENV

    afterAll(() => {
      process.env.NODE_ENV = origNodeEnv
    })

    it('should wrap display name in non-production env', () => {
      process.env.NODE_ENV = 'test'

      const EnhancedTarget = withCallbackOnce()(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })

    it('should not wrap display name in production env', () => {
      process.env.NODE_ENV = 'production'

      const EnhancedTarget = withCallbackOnce()(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
