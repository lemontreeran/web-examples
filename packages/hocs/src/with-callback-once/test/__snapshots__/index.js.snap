// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withCallbackOnce display name should not wrap display name in production env 1`] = `
<WithCallbackOnce>
  <Target />
</WithCallbackOnce>
`;

exports[`withCallbackOnce display name should wrap display name in non-production env 1`] = `
<withCallbackOnce(Target)>
  <Target />
</withCallbackOnce(Target)>
`;

exports[`withCallbackOnce should invoke a callback when condition status has been changed 1`] = `Array []`;

exports[`withCallbackOnce should invoke a callback when condition status has been changed 2`] = `
Array [
  Array [
    Object {
      "a": 11,
      "b": 2,
      "c": 3,
    },
  ],
]
`;

exports[`withCallbackOnce should no-op if condition status is the same 1`] = `Array []`;

exports[`withCallbackOnce should no-op if condition status is the same 2`] = `Array []`;

exports[`withCallbackOnce should pass props through 1`] = `
<withCallbackOnce(Target)
  a={1}
  b={2}
  c={3}
>
  <Target
    a={1}
    b={2}
    c={3}
  />
</withCallbackOnce(Target)>
`;
